1-raw-data/mailinglists/googlegroups/googlegroups_data.go

```diff
@@ -0,0 +1,416 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/*
+This package loads Google Groups mailing list data types into Cloud Storage.
+
+List of all topics / threads url format| LIFO:
+topicListURLBase = "https://groups.google.com%s/forum/?_escaped_fragment_=forum/%s"
+https://groups.google.com/forum/?_escaped_fragment_=forum/[GROUP NAME]
+Pass in numbers [1-100] to get list of results
+- Note keep between 1-100 in size and grow accordingly 10-110, 201-300, etc
+- Link only to topic and not post
+
+List of all messages in a topic url format | FIFO:
+topicDetailURLBase = "https://groups.google.com%s/forum/?_escaped_fragment_=topic/%s/%s"
+https://groups.google.com/forum/?_escaped_fragment_=topic/[GROUP NAME]/[TOPIC ID]
+
+Topic pages contain results in this format with message id which we want:
+"https://groups.google.com/d/msg/%s/%s/%s" // group, topic, msg id
+
+Raw mail message url format:
+messageURLBase = "https://groups.google.com%s/forum/message/raw?msg=%s/%s/%s"
+https://groups.google.com/forum/message/raw?msg=[GROUP NAME]/[TOPIC ID]/[MSG ID]
+
+Atom links:
+- Gets similar information where msgs focus on msg content and topics pulls labeled sections like Approvals
+https://groups.google.com/forum/feed/[GROUP NAME]/msgs/atom.xml?num=100
+https://groups.google.com/forum/feed/[GROUP NAME]/topics/atom.xml?num=50
+
+RSS link with link to msg and topic:
+https://groups.google.com/forum/feed/[GROUP NAME]/msgs/rss.xml?num=50
+https://groups.google.com/forum/feed/[GROUP NAME]/topics/rss.xml?num=50
+
+Note Ajax crawling and escaped_fragment is deprecated and this will need to be revised to align to current approaches
+
+*/
+
+package googlegroups
+
+import (
+	"context"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"net/http"
+	"path"
+	"regexp"
+	"sort"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/PuerkitoBio/goquery"
+	"github.com/google/project-OCEAN/1-raw-data/gcs"
+)
+
+// TODO setup gobal errors to pass and test
+// TODO setup so can pull specific dates
+// TODO setup to better handle different http errors and capturing where it fails
+
+type Results struct {
+	urlMap map[string][]string
+	err    error
+}
+
+type jobsData struct {
+	topicURLList []string
+	fileName     string
+}
+
+func httpStringResponse(url string) (responseString string, err error) {
+	var (
+		bodyBytes []byte
+		response  *http.Response
+	)
+
+	if url == "" {
+		return
```

Comment:
You need to return two items here.

Reply:
Same for the other return statements.

Reply:
Because I've labeled the return values in the header it defaults to an empty string and nil which is what I want. I'll add a comment here but this is to keep the program running even when the url is empty.